### PR TITLE
Add an LRU cache to util library

### PR DESCRIPTION
**Pull request by andrewjstone**

### CODE REVIEW

File: util/include/lru_cache.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    return keys_.size();
+  }
+
+  struct Stats {
+    size_t hits = 0;
+    size_t misses = 0;
+    size_t puts = 0;
+  };
+
+  Stats getStats() { return stats_; }
```
Comment (dartdart26) [inline util/include/lru_cache.hpp:69]: Can be made `const`.

File: util/include/lru_cache.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+ private:
+  size_t capacity_;
```
Comment (dartdart26) [inline util/include/lru_cache.hpp:72]: Can be made `const`.

File: util/include/lru_cache.hpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        auto& key = keys_.back();
+        map_.erase(key);
+        keys_.pop_back();
+      }
+      keys_.push_front(key);
+      map_.insert({key, {value, keys_.begin()}});
+    }
+  }
+
+  std::optional<Value> get(const Key& key) {
```
[the remark this answers is not in the document]
Comment (dartdart26) [reply util/include/lru_cache.hpp:46]: Actually, now I am not sure about it as calls to `get()` can be observed by another `get()` calls after calls to` put()`. Maybe non-const is the right thing.

**Review by dartdart26 — APPROVED** [review]
Looks great, thanks @andrewjstone! I've added minor comments only that you can decide whether to address or not.

**Comment by andrewjstone** [comment]
Thanks for the great review @dartdart26. I'm taking you up on all your suggestions, except the const get :)

File: util/include/lru_cache.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply util/include/lru_cache.hpp:46]: I had it const originally, but changed it because it does make user visible changes.

**Pull request merged**